src/gadgetio.cxx

```diff
@@ -172,20 +172,21 @@ void ReadGadget(Options &opt, vector<Particle> &Part, const Int_t nbodies,Partic
         opt.Omega_Lambda=header[ifirstfile].OmegaLambda;
         opt.h=header[ifirstfile].HubbleParam;
         opt.Omega_cdm=opt.Omega_m-opt.Omega_b;
+        CalcOmegak(opt);
         //Hubble flow
         if (opt.comove) aadjust=1.0;
         else aadjust=opt.a;
-        Hubble=opt.h*opt.H*sqrt((1-opt.Omega_m-opt.Omega_Lambda)*pow(aadjust,-2.0)+opt.Omega_m*pow(aadjust,-3.0)+opt.Omega_Lambda);
-        opt.rhobg=3.*Hubble*Hubble/(8.0*M_PI*opt.G)*opt.Omega_m;
-        Double_t bnx=-((1-opt.Omega_m-opt.Omega_Lambda)*pow(aadjust,-2.0)+opt.Omega_Lambda)/((1-opt.Omega_m-opt.Omega_Lambda)*pow(aadjust,-2.0)+opt.Omega_m*pow(aadjust,-3.0)+opt.Omega_Lambda);
-        opt.virBN98=(18.0*M_PI*M_PI+82.0*bnx-39*bnx*bnx)/opt.Omega_m;
+        Hubble=GetHubble(opt, aadjust);
+        CalcCriticalDensity(opt, aadjust);
+        CalcBackgroundDensity(opt, aadjust);
+        CalcVirBN98(opt,aadjust);
         //if opt.virlevel<0, then use virial overdensity based on Bryan and Norman 1997 virialization level is given by
         if (opt.virlevel<0) opt.virlevel=opt.virBN98;
 
         //normally Hubbleflow=lvscale*Hubble but we only care about peculiar velocities
         //ignore hubble flow
         Hubbleflow=0.;
-        cout<<"Cosmology (h,Omega_m,Omega_cdm,Omega_b,Omega_L) = ("<< opt.h<<","<<opt.Omega_m<<","<<opt.Omega_cdm<<","<<opt.Omega_b<<","<<opt.Omega_Lambda<<")"<<endl;
+        PrintCosmology(opt);
```

Comment:
The same calculations done here are done in all data input routines, including the swift interface: Omega-k, critical density, background density, and VirBN98. All these are also done in `CalcCosmoParams`, which is defined in `substructureproperties.cxx` but neither declared nor used. On top of that there is more common code to all the data input routines: the calculation of the adjustment factor from the comove option, the adjustment of the virlevel option when it's negative, and the printing of the cosmological parameters.

Taking this into account, I'd take this opportunity to move these three additional operations into `CalcCosmoParams` (which then wouldn't take an extra Double_t argument, as it would be calculated internally), and then add a call to `CalcCosmoParams` from each data input routine replacing the current common contents.

As a result, the calls to GetHubble inside each data input routine are not necessary anymore, and the Hubble variable (and probably the aadjust variable) can be removed.

Reply:
Just to clarify: the code displayed above doesn't give the full context for the comment, which is better seen in the "Files changed" tab of the pull request.